util/merkle-tree/src/merkle_tree.rs

```diff
@@ -0,0 +1,354 @@
+
+    while queue.len() > 1 {
+        let right = queue.pop_front().unwrap();
+        let left = queue.pop_front().unwrap();
+        queue.push_back(merge(&left, &right));
+    }
+
+    queue.pop_front().unwrap()
+}
+
+pub fn build_merkle_tree(leaves: Vec<H256>) -> Vec<H256> {
+    if leaves.is_empty() {
+        return vec![];
+    }
+
+    let mut nodes = vec![H256::zero(); leaves.len() - 1];
+    nodes.extend(leaves);
+    let mut i = nodes.len() - 1;
+
+    while i != 0 {
+        nodes[(i - 1) >> 1] = merge(&nodes[i - 1], &nodes[i]);
+        i -= 2;
+    }
+
+    nodes
+}
+
+pub fn update_root(tree: &mut Vec<H256>, mut index: usize) {
+    if index >= tree.len() {
+        return;
+    }
+
+    while index != 0 {
+        let sibling = ((index + 1) ^ 1) - 1;
+        let tmp = (index - 1) >> 1;
+        tree[tmp] = if sibling > index {
+            merge(&tree[index], &tree[sibling])
+        } else {
+            merge(&tree[sibling], &tree[index])
+        };
+        index = tmp;
+    }
+}
+
+#[derive(Eq, PartialEq, Debug, Serialize, Deserialize, Clone)]
+pub struct Proof {
+    // From 0
+    pub index: u32,
+    pub nodes: Vec<H256>,
+}
+
+impl Proof {
+    pub fn root(&self, mut hash: H256) -> Option<H256> {
+        let mut index = self.index;
+        let len = 32 - (index + 1).leading_zeros() - 1;
+
+        if self.nodes.len() != len as usize {
+            return None;
+        }
+
+        hash = self.nodes.iter().fold(hash, |x1, x2| {
+            let ret = if index & 1 == 0 {
+                merge(x2, &x1)
+            } else {
+                merge(&x1, x2)
+            };
+            index = (index - 1) >> 1;
+            ret
+        });
+
+        Some(hash)
+    }
+
+    pub fn verify(&self, root: &H256, hash: H256) -> bool {
+        if let Some(x) = self.root(hash) {
+            &x == root
+        } else {
+            false
+        }
+    }
+}
+
+pub fn build_proof(tree: &[H256], index: usize) -> Option<Proof> {
+    if index >= tree.len() {
+        return None;
+    }
+    let mut nodes = Vec::new();
+    let mut curr = index;
+
+    while curr != 0 {
+        let sibling = ((curr + 1) ^ 1) - 1;
+        nodes.push(tree[sibling].clone());
+        curr = (curr - 1) >> 1;
+    }
+    let index = index as u32;
+    Some(Proof { index, nodes })
+}
+
+pub fn calc_sibling(num: usize) -> usize {
+    if num == 0 {
+        0
+    } else {
+        ((num + 1) ^ 1) - 1
+    }
+}
+
+pub fn calc_parent(num: usize) -> usize {
+    if num == 0 {
+        0
+    } else {
+        (num - 1) >> 1
+    }
+}
+
+pub fn is_left(num: usize) -> bool {
+    num & 1 == 1
+}
+
+#[derive(Eq, PartialEq, Debug, Serialize, Deserialize, Clone)]
+pub struct CombinedProof {
+    // size of items in the tree
+    pub size: u32,
+    // nodes on the path which can not be calculated, in descending order by index
+    pub nodes: Vec<H256>,
+}
+
+impl CombinedProof {
+    pub fn root(&self, mut tuples: Vec<(u32, H256)>) -> Option<H256> {
```

Comment:
This function contains much nonintuitive code, I need more time to understand it, will review it later after we unify the `CombinedProof` and `Proof` to single struct